kube-runtime/src/watcher.rs

```diff
@@ -225,15 +233,28 @@ async fn step<K: Resource + Clone + DeserializeOwned + Debug + Send + 'static>(
 /// that we have seen on the stream. If this is successful then the stream is simply resumed from where it left off.
 /// If this fails because the resource version is no longer valid then we start over with a new stream, starting with
 /// an [`Event::Restarted`].
-pub fn watcher<K: Resource + Clone + DeserializeOwned + Debug + Send + 'static>(
+pub fn watcher<K: Resource + Clone + DeserializeOwned + Debug + Send + 'static, B: Backoff + Send>(
     api: Api<K>,
     list_params: ListParams,
+    backoff: B,
 ) -> impl Stream<Item = Result<Event<K>>> + Send {
     futures::stream::unfold(
-        (api, list_params, State::Empty),
-        |(api, list_params, state)| async {
+        (api, list_params, backoff, State::Empty),
+        |(api, list_params, mut backoff, state)| async {
             let (event, state) = step(&api, &list_params, state).await;
-            Some((event, (api, list_params, state)))
+            if event.is_err() {
+                if let Some(wait_time) = backoff.next_backoff() {
+                    tracing::debug!("watch waiting {}ms until retrying", wait_time.as_millis()); // TODO: kind name here
```

Comment:
Still need to bubble up the original error here. Currently `BackoffRetriable` is the only error type that you ever return.

Reply:
yeah, that was because we needed to catch it later to avoid it bubbling up (and couldn't wrap the original error inside the `BackoffRetriable` because that would be a recursive error). Going to pursue the utility approach you suggested for doing this logic instead and having an error in there. That way `watcher` can bubble up everything, but the utility can do something like this, and the `Observer` can interface with it.

Reply:
the error type is gone now. but nothing is bubbled up, the point is that what is bubbled up is crash worthy.